geps/gep-2643/index.md

```diff
@@ -0,0 +1,263 @@
+* Support attributes other than the SNI/hostname for the route selection
+
+## Introduction
+
+While many application routing cases can be implemented using HTTP/L7 matching
+(the tuple protocol:hostname:port:path), there are some specific cases where direct,
+encrypted communication to the backend may be required, without further assertion.  For example:
+* A backend that is TLS based but not HTTP based (e.g., a Kafka service, or a
+Postgres service, with its listener being TLS enabled).
+* Some WebRTC solutions.
+* Backends that can require direct client-certificate authentication (e.g., OAuth).
+
+For the example cases above, it is desired that the routing is made as a passthrough
+mode, where the `Gateway` passes the packets to the backend without terminating TLS.
+
+While this routing is possible using other mechanisms such as a `TCPRoute` or a
+Kubernetes service of type `LoadBalancer`, it may be desired to have a single point
+of traffic convergence (like one single `LoadBalancer`) for reasons like cost,
+traffic control, etc.
+
+An implementation of `TLSRoute` achieves this end using the
+[server_name TLS attribute](https://datatracker.ietf.org/doc/html/rfc6066#section-3)
+to determine what backend should be used for a given request. `TLSRoute` thereby
+enables the use of a single gateway listener to handle traffic for multiple routes.
+
+## API
+
+```golang
+type TLSRouteSpec struct {
+  // Hostnames defines a set of SNI hostnames that should match against the
+  // SNI attribute of TLS ClientHello message in TLS handshake. This matches
+  // the RFC 1123 definition of a hostname with 2 notable exceptions:
+  //
+  // 1. IPs are not allowed in SNI hostnames per RFC 6066.
+  // 2. A hostname may be prefixed with a wildcard label (`*.`). The wildcard
+  //    label must appear by itself as the first label.
+  //
+  // If a hostname is specified by both the Listener and TLSRoute, there
+  // must be at least one intersecting hostname for the TLSRoute to be
+  // attached to the Listener. For example:
+  //
+  // * A Listener with `test.example.com` as the hostname matches TLSRoutes
+  //   that have specified at least one of `test.example.com` or
+  //   `*.example.com`.
+  // * A Listener with `*.example.com` as the hostname matches TLSRoutes
+  //   that have specified at least one hostname that matches the Listener
+  //   hostname. For example, `test.example.com` and `*.example.com` would both
+  //   match. On the other hand, `example.com` and `test.example.net` would not
+  //   match.
+  //
+  // If both the Listener and TLSRoute have specified hostnames, any
+  // TLSRoute hostnames that do not match the Listener hostname MUST be
+  // ignored. For example, if a Listener specified `*.example.com`, and the
+  // TLSRoute specified `test.example.com` and `test.example.net`,
+  // `test.example.net` must not be considered for a match.
+  //
+  // If both the Listener and TLSRoute have specified hostnames, and none
```

Comment:
I think it would be good to precise that Listener's protocol needs to be of type: TLS and that TLSRoutes only applies to Listeners with this protocol. Is this said somewhere in the doc/spec?

Reply:
agreed. As a heads up, the API here was copied exactly from the existing code, so the idea is that we also make the proper updates on the API based on the comments here.

That said, we do not explicitly say anywhere here that TLSRoute is attacheable to Listeners of type TLS and Passthrough, just on places like https://gateway-api.sigs.k8s.io/guides/tls/

I will make this explicit on this doc

Reply:
Added as part of TLSRouteSpec, closing this